File: 03_03_DinnerPlates.java
class DinnerPlates {
    
    int maxCapacity;
    
    List<Deque<Integer>> stacks = new ArrayList<>();
    TreeSet<Integer> treeSet = new TreeSet<>();

    public DinnerPlates(int capacity) {
        this.maxCapacity = capacity;
    }
    
    public void push(int val) {
        
        if (!treeSet.isEmpty()) {
            int lowestIndex = treeSet.pollFirst();
            
            if (lowestIndex == stacks.size()) {
                stacks.add(new ArrayDeque<>());
            }
            
            Deque<Integer> nonEmptyStack = stacks.get(lowestIndex);
            nonEmptyStack.push(val);
            if (nonEmptyStack.size() < maxCapacity) {
                treeSet.add(lowestIndex);
            }
            return;
        }
        
        
       
        Deque<Integer> newStack = new ArrayDeque<>();
        newStack.push(val);
        
        if (maxCapacity != 1) {
            treeSet.add(stacks.size());
        }
        
        stacks.add(newStack);
        
    }
    
    public int pop() {
        if (stacks.isEmpty()) {
            return -1;
        } else {
            Deque<Integer> currentStack = stacks.get(stacks.size() - 1);
            
            while (currentStack.isEmpty()) {
                stacks.remove(stacks.size() - 1);
                if (stacks.isEmpty()) {
                    return -1;
                } else {
                    currentStack = stacks.get(stacks.size() - 1);
                }
            }
            treeSet.add(stacks.size() - 1);
            return currentStack.pop();
        }
    }
    
    public int popAtStack(int index) {
        if (index >= stacks.size()) {
            return -1;
        }
        
        Deque<Integer> stack = stacks.get(index);
        
        if (stack.isEmpty()) {
            return -1;
        } else {
            treeSet.add(index);
            return stack.pop();
        }  
    }
}

/**
 * Your DinnerPlates object will be instantiated and called as such:
 * DinnerPlates obj = new DinnerPlates(capacity);
 * obj.push(val);
 * int param_2 = obj.pop();
 * int param_3 = obj.popAtStack(index);
 */

File: 03_03_DinnerPlates.cpp
class DinnerPlates {
    public:
    int capacity;
    set<int> availableStacks;
    vector<vector<int>> stacks;
    
    DinnerPlates(int capacity) {
        this->capacity = capacity;
        availableStacks.clear();
        stacks.clear();
    }
    
    void push(int val) {

        if (availableStacks.size() > 0)
        {
            auto minIdx = *availableStacks.begin();
            availableStacks.erase(minIdx);
            
            if (minIdx == stacks.size())
                stacks.push_back(vector<int>());

            auto & current = stacks.at(minIdx);
            current.push_back(val);
            
            if(current.size() < this->capacity)
                availableStacks.insert(minIdx);
        }
        else
        {
            auto current = vector<int>(1, val);
            
            if (current.size() < this->capacity)
                availableStacks.insert(stacks.size());

            stacks.push_back(current);
        }
    }
    
    int pop() {
        if (stacks.size() == 0)
            return -1;
        
        auto * currentStack = &stacks.at(stacks.size() - 1);
        
        while(currentStack->size() == 0)
        {
            stacks.pop_back();
            if(stacks.size() == 0)
                return -1;
            currentStack = &stacks.at(stacks.size() - 1);
        }
    
        availableStacks.insert(stacks.size() - 1);
        auto poppedValue = currentStack->at(currentStack->size() - 1); 
        currentStack->pop_back();

        return poppedValue;      
    }
    
    int popAtStack(int index) {

        if (index >= stacks.size())
            return -1;
        
        auto & current = stacks.at(index);

        if (current.size() == 0)
            return -1;

        availableStacks.insert(index);
        auto poppedValue = current[current.size() - 1]; current.pop_back();

        return poppedValue;
    }
};
